io/jvm/src/main/scala/scalaz/effect/RTS.scala

```diff
@@ -463,20 +462,22 @@ private object RTS {
                       }
                     } else {
                       // Error caught:
-                      val value = -\/(error)
+                      // FIXME this cast is not safe, `err` is defined as `E => IO[E2, B]`
+                      val handledIo = errorHandler(error).asInstanceOf[IO[E, Any]]
```

Comment:
I think it actually is but I'll spend a moment on it.

Reply:
The new `attempt` is sort of a `flatMap` it allows the user to create any sort of `IO` as a response to `E` or `A`, they can even hit the DB to handle any of the branches, so we can expect the error in the resultant `IO` to be of any type. That's why `attempt` introduces a new type parameter in first place.

Reply:
But remember, we are running this from the "top", the fully composed `IO[E, A]`. So the `E` that you see in lower level are NOT the same as the `E` at the top. In fact, we should probably only use `E` for a top-level error (although I hate to introduce `Any` more places). The user, in handling an error, is bringing up a lower-level `E` into the `E` used at the top computation.

The type system cannot describe this so we have to coerce at some place. The question is where is the best place to coerce that preserves the most type soundness.

Reply:
Definition on this? have you given it any more thoughts?

Reply:
It's safe. I still am not sure how to improve type safety but I may look at that myself after the current ticket is done.